Adds support using temporary security credentials

d2r opened the pull request:
https://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html

d2r commented:
Looking at build issues.

d2r commented on plugins/s3_auth/unit_tests/test_aws_auth_v4.cc:
```diff
@@ -675,7 +678,42 @@ TEST_CASE("S3AuthV4UtilParams: include all headers by default", "[AWS][auth][uti
+  };
+
+  ValidateBenchCanonicalRequest(api, /*signPayload */ false, &now, bench, include, exclude);
 }
```
This duplicates the previous test, but adds the new header field.

d2r replied:
* Remove the added test, because the signing of arbitrary headers is already covered by existing tests.
* Add a test to confirm that the new header is added when the plugin is configured to add it.

Do I have that right?

gtenev replied:
> Remove the added test, because the signing of arbitrary headers is already covered by existing tests.

right.  

> Add a test to confirm that the new header is added when the plugin is configured to add it.

The Catch unit tests test only the v4 calculation. Your changes are in the plugin which will be hard to test from the unit-test we have. I guess you would need to create AuTest test which you will have to be created from scratch for this plugin (it will be great but it is up to you).

d2r commented on plugins/s3_auth/unit_tests/test_aws_auth_v4.cc:
```diff
@@ -286,18 +286,24 @@ TEST_CASE("AWSRegions: get region by providing bucket name having multiple label
+  "bg9zErVSjEXAMPLE//////////TnBjaVKC0y/9Y3zSyB8FtGpsu5Lf83k1exfP2VEereNnx1SxaFtzxk7f61/Eq81ud4Opuo/tslm5xrufFOKa34SvDfBaBl5r1V73/"
+  "7APNC4E0f0tK4u+"
+  "uOoqyq32YXJX2yNJ0K4Ud3MLwH9LfFgDkdpFW0J4eh0Ag8kxURpqHVOWlnIzSXHuDIZwx5Oq9WiWpNJqyEPPYNgjAkKVgvdf7g56tnfkuhyMFt1LN8xwOt2dPgiUjDhe"
+  "x/3O+1LqlAPUQYzzpcn8Mqgj5tvsDz0wTMtUbTQov1hFKGuTMxScD5m16QZv3A7vSde1aGa9Vgckg5c2TU91dQdtDodmPUpMZ24HZ11oBgv0PpREXAMPLE";
```
This is what `clang-format` wants.

d2r replied:
I'll try it and see.

d2r commented on plugins/s3_auth/aws_auth_v4.h:
```diff
@@ -101,6 +103,8 @@ class AwsAuthV4
   const char *_awsSecretAccessKey = nullptr;
   size_t _awsSecretAccessKeyLen   = 0;
+  const char *_awsSessionToken    = nullptr;
+  size_t _awsSessionTokenLen      = 0;
```
This class may not need to store the token itself.

d2r commented on plugins/s3_auth/unit_tests/test_aws_auth_v4.cc:
```diff
@@ -392,7 +397,7 @@ TEST_CASE("AWSAuthSpecByExample: GET Object", "[AWS][auth][SpecByExample]")
   };
 
-  ValidateBench(api, /*signePayload */ true, &now, bench, defaultIncludeHeaders, defaultExcludeHeaders);
+  ValidateBench(api, /*signPayload */ true, &now, bench, defaultIncludeHeaders, defaultExcludeHeaders);
```
I am willing to split these code comment changes into a separate PR if that's what we want.

gtenev commented on plugins/s3_auth/s3_auth.cc:
```diff
@@ -200,6 +201,9 @@ class S3Config
         TSError("[%s] region map is not used with AWS auth v2, parameter ignored", PLUGIN_NAME);
       }
+      if (nullptr != _token || _token_len > 0) {
+        TSError("[%s] session token is not used with AWS auth v2, parameter ignored", PLUGIN_NAME);
```
> I need to either clarify error messages and the documentation here or else make sure this works with v2 as well.

I think v2 is probably going out of life soon so may be you could just clarify the docs and the error message (just to make sure it is clear to the operator that v2 is not supported)

d2r commented:
OK, sorry this took so long. I addressed review comments:
* Squashed commits, old commit was a0d9df315897da802ffac9ad0a8c54075cd8834c
* Opted not to add the functional test
* Updated documentation to be consistent

d2r commented:
@gtenev ping

gtenev commented on doc/admin-guide/plugins/s3_auth.en.rst:
```diff
@@ -36,23 +36,25 @@ Using the plugin in a remap rule would be e.g.::
-    version=2
+    session_token=my-token
+    version=4
     virtual_host=yes
```
`virtual_host` is a v2 specific parameter. We have 2 options here:

1. remove `virtual_host` from this example (otherwise plugin will warn in the log on each realod/restart)

1. change the example back to v2 like it was before (you have already added a good example in the v4 specific section) and remove `session_token=my-token`

d2r replied:
Thanks, opting for 2.

d2r commented:
@gtenev Addressed review comment.

bryancall commented:
Cherry picked to 8.1.0
